### ISSUE TITLE
Impeller drawVertices color blending differs from Skia

### ISSUE THREAD

**Issue by solgar** [flutter/flutter#143363]
### Steps to reproduce

1. Create CustomPaint with CustomPainter
2. Create `Vertices` with colors
3. Paint created vertices using `drawVertices` with **any** `blendMode` and with `Paint()` with transparent color and default blend mode (default is `BlendMode.srcOver`)
4. Compare differences between Skia and Impeller renderers

Run attached code using Skia on macOS or iOS (haven't checked on Android), then run it again using Impeller.

Changing between Skia and Impeller can be done by editing `Info.plist`:
```
<key>FLTEnableImpeller</key>
<true />
```

### Expected results

Assuming Skia has correct behavior:
<img width="462" alt="skia" src="https://github.com/flutter/flutter/assets/312721/1c70f875-7790-4efa-aee5-2a98761fe476">

### Actual results

Assuming Impeller has incorrect behavior:
<img width="462" alt="impeller" src="https://github.com/flutter/flutter/assets/312721/abe41652-ce8a-4cf2-9678-221d0c393d3a">

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'dart:ui';

import 'package:flutter/material.dart';

void main() {
  runApp(CustomPaint(
    painter: DrawVerticesBugPainter(),
    child: const SizedBox.expand(),
  ));
}

class DrawVerticesBugPainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    // ignore: prefer_const_declarations
    final margin = 64.0;
    canvas.drawColor(Colors.white, BlendMode.src);
    canvas.drawLine(
        Offset.zero,
        Offset(size.width, size.height),
        Paint()
          ..color = Colors.black
          ..strokeWidth = 10);

    final vertices = Vertices(
        VertexMode.triangles,
        [
          const Offset(0, 0),
          Offset(0, size.height - margin * 2),
          Offset(size.width - margin * 2, 0),
          Offset(size.width - margin * 2, size.height - margin * 2)
        ],
        indices: [0, 1, 2, 1, 2, 3],
        colors: List.filled(4, Colors.red));

    canvas.translate(margin, margin);
    canvas.drawVertices(
      vertices,
      BlendMode.dst,
      Paint()..color = Colors.black.withAlpha(200),
    );
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>
Another example of Skia:
<img width="766" alt="skia2" src="https://github.com/flutter/flutter/assets/312721/836f4ce6-df80-4005-a415-1da8b6fbfbdb">

Another example of Impeller:
<img width="766" alt="impeller2" src="https://github.com/flutter/flutter/assets/312721/5672bc08-4f06-434c-814d-e37f88567f40">

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.16.9, on macOS 14.1.1 23B81 darwin-arm64, locale en-PL)
    • Flutter version 3.16.9 on channel stable at /Users/igor/fvm/versions/3.16.9
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 41456452f2 (3 weeks ago), 2024-01-25 10:06:23 -0800
    • Engine revision f40e976bed
    • Dart version 3.2.6
    • DevTools version 2.28.5

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/igor/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/igor/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode-15.2.0.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Users/igor/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Community Edition (version 2023.2.2)
    • IntelliJ at /Users/igor/Applications/IntelliJ IDEA Community Edition.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.86.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.82.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 14.1.1 23B81 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 121.0.6167.160

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by solgar** [comment]
This issue is similar to or maybe even duplicate of https://github.com/flutter/flutter/issues/142549

**Comment by jonahwilliams** [comment]
The opacity issue has been fixed as of 3.19. I'm unsure of the exact commit, though its unlikely that we will cherry pick it due to the impending next stable.

The remaining issue is that solid colors (on the paint) are special cased in both backends, and the documentation needs to be updated to clarify this.

**Comment by solgar** [comment]
Checked on `3.19.0-0.4.pre` and macOS Impeller renders same as Skia, which is good. There is still an issue with documentation, but this one is really about rendering itself. Closing. Thanks @jonahwilliams !
